spec/message_spec.rb

```diff
@@ -0,0 +1,67 @@
+RSpec.describe "message encoding" do
+  # message Sally {
+  #   required string mood = 1;
+  #   optional uint32 age  = 2;
+  # }
+  class Sally
+    def initialize(attributes)
+      @attributes = {}
+      attributes.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+    end
+
+    ## Attributes
+
+    def age
+      @attributes[:age]
+    end
+
+    def age=(age)
+      raise ArgumentError, "age is not a number" unless age.respond_to?(:to_i)
+      raise ArgumentError, "age is too large" unless age.to_i <= 4_294_967_295
```

Comment:
I wish we could do the bounds checking here and do the coercion later, but we can't check bounds until after coercion, so maybe we just defer the bounds check until encoding takes place?